### PR TITLE
bug in combine

### PR DESCRIPTION
**Pull request by abuts**
Fixes Re #1778.

Combine experiments have been completely refactored to properly account for duplicated headers and other experiment components.

Bugs, identified by user are fixed. 
Also fixed old code with `combine_dnd` though some unification between old and new code is necessary.
Changed treatment of duplicated runs. While previously they received fake ID, now they contribute to duplicated run_id. `IX_experiment` values for runs with the same `run_id` have to be the same.

Needed some reconciliation beteween old `d1d, d2d` combine, new `sqw` combine and cut with non-aligned axis -- they use absolutely different approaches for the same purpose.

### CODE REVIEW

File: herbert_core/utilities/general/replace_inf_range.m [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+%   inf, 1,   inf, 20]
+% The output would be
+%  [-1    ,1-eps,-eps, 10,...
+%   -1+eps,1    , eps, 20]
+% Where eps is smallest positive compatible with 1 i.e. 1+eps ~= 1
+%
+% Horace-4 does not support inf ranges any more.
+% Unlike previous Horace version, where  -inf:inf mean whole integration
+% range, dnd with 0-integration range is meaningless for dnd obj,
+% so will set +-0
```
Comment (cmarooney-stfc) [inline herbert_core/utilities/general/replace_inf_range.m:22]: Not clear about this. The old way had -inf:inf as the whole integration range. What is the representation of the whole integration range in this new method? Using -eps for -inf in the above example suggests that negative ranges are not possible. Is this the case?

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+%                    and pointing to the nuber of element in obj
+%                    as value.
+%
+% Returns:
+% obj             -- resulting array, containing unique
+%                    instances of IX_experiment classes with
+%                    all non-unique IX_experiments excluded.
+% skipped_inputs  -- cellarray of logical arrays containing true where
+%                    input object was dropped and false where it has been
+%                    kept
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m:30]: well, if you provide two identical arrays of `IX_experiment`s, combine would not duplicate them, and return only unique copies in the outiput array. 

this cellarray will contain information about the fact that some inputs were dropped. I do not know how to describe this better, though have expanded the explanation. May be insufficiently.

File: herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+% file_id_array   -- array contains run_ids for each input IX_experiment value.
+%                    if some IX_experiments with equal run_id(s) and parameters
+% values
+%                    were rejected, some elements of this array will
+%                    contain the same values to be used in calculations of
+%                    pixels run_id for each contributed file.
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m:36]: again, I've expanded explanation, but it relates to the details of the algorithm, so it is diffucult to understand without understanding the algorithm itself.

File: herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    if ~keep_runid
+        [obj,this_runid_map,file_id_array] = recalc_runid(obj,this_runid_map,file_id_array);
+    end
+
+    skipped_inputs = {};
+    return;
+end
+
+if isa(exper_cellarray{1},'Experiment')
+    exper_cellarray = cellfun(@(x)(x.expdata),exper_cellarray,'UniformOutput',false);
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/@IX_experiment/private/combine_.m:57]: I've added some explanation, but usually commenting what code is doing is a bad practice. You see it from the code itself. Proper comment should highlight intentions.

File: herbert_core/utilities/general/replace_inf_range.m (hunk 1 above)
Comment (abuts) [reply herbert_core/utilities/general/replace_inf_range.m:22]: hopefully clarified

**Pull request merged**